prov/shm/src/smr_msg.c

```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
 		uint64_t tag, uint64_t ignore, uint64_t flags,
 		struct smr_queue *recv_queue, struct smr_queue *unexp_queue)
 {
-	struct smr_rx_entry *entry;
-	ssize_t ret = -FI_EAGAIN;
-	fi_addr_t use_addr;
+	struct smr_match_attr match_attr;
+	struct smr_rx_entry *unexp, *owner_entry = NULL;
```

Comment:
This function is hard to follow.  I think adding an *mrecv_entry and using that instead of owner_entry would help.  The code has a lot of checks around owner_entry being non-null, but what it is actually trying to check is if the posted receive is a multi-recv entry.  owner_entry is used for both multi-recv and normal receives, but switches its meaning 3/4th of the way through the function after the while loop.